### PR TITLE
KAFKA-14274, #7: introduce FetchRequestManager to integrate fetch into new consumer threading refactor

### PR DESCRIPTION
**Pull request by kirktrue**
This change introduces the `FetchRequestManager` that will be responsible for:

- Formatting fetch requests to the background thread
- Configuring the callback on fetch responses for the background thread

The response handler will collect the fetch responses from the broker and create `CompletedFetch` instances as is done in `Fetcher`. However, unlike the `Fetcher`, the foreground logic will decompress/deserialize/and convert the data in `CompletedFetch` before returning to the user.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java [hunk 1]
```diff
@@ -458,8 +312,8 @@ protected Map<Node, FetchSessionHandler.FetchRequestData> prepareFetchRequests()
             // Use the preferred read replica if set, otherwise the partition's leader
             Node node = selectReadReplica(partition, leaderOpt.get(), currentTimeMs);
 
-            if (client.isUnavailable(node)) {
-                client.maybeThrowAuthFailure(node);
+            if (nodeStatusDetector.isUnavailable(node)) {
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java]: Just for the record, if we agree on this NodeStatusDetector component, I would integrate a similar logic to the ListOffsetRequestManager for checking leader status before sending a request. lgtm

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java [hunk 2]
```diff
@@ -33,36 +38,92 @@
...
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.retryBackoffMs = retryBackoffMs;
         this.client = client;
-        this.log = logContext.logger(getClass());
         this.unsentRequests = new ArrayDeque<>();
-        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
-        this.retryBackoffMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+    }
+
+    public static NetworkClientDelegate create(final LogContext logContext,
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: We are introducing this static factory method but keeping the public constructors with rather similar set of arguments, which can be confusing I would say. Should we maybe provide a consistent way of creating a new client (I do like the factory method here, although seems not used anywhere yet btw)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java [hunk 3]
```diff
@@ -33,36 +38,92 @@
...
-            final LogContext logContext,
-            final KafkaClient client) {
+    private final ConcurrentLinkedQueue<Node> pendingDisconnects = new ConcurrentLinkedQueue<>();
+
+    public NetworkClientDelegate(final LogContext logContext,
+                                 final Time time,
+                                 final ConsumerConfig config,
+                                 final KafkaClient client) {
+        this(logContext,
+                time,
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: Indentation not right

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.KafkaClient;
+import org.apache.kafka.clients.NetworkClientUtils;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.errors.AuthenticationException;
+import org.apache.kafka.common.utils.Time;
+
+/**
+ * A user of {@code NodeStatusDetector} can determine the status of a given broker {@link Node} to determine
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java]: Review "can determine...to determine..."

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * its status. It's also possible to check for previous authentication errors if the node isn't available.
+ *
+ * @see ConsumerNetworkClient
+ * @see NetworkClientDelegate
+ */
+public interface NodeStatusDetector {
+
+    /**
+     * Check if the code is disconnected and unavailable for immediate reconnection (i.e. if it is in
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java]: nit: typo "code" vs "node"

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * reconnect backoff window following the disconnect).
+     *
+     * @param node {@link Node} to check for availability
+     * @see NetworkClientUtils#isUnavailable(KafkaClient, Node, Time)
+     */
+    boolean isUnavailable(Node node);
+
+    /**
+     * Check if the code is disconnected and unavailable for immediate reconnection (i.e. if it is in
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java]: Not the right description for `maybeThrowAuthenticationException`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java [hunk 7]
```diff
@@ -113,54 +132,99 @@ public PrototypeAsyncConsumer(final Map<String, Object> configs,
...
-                           Metrics metrics,
-                           Optional<String> groupId,
-                           int defaultApiTimeoutMs) {
-        this.time = time;
+                metrics,
+                subscriptions,
+                groupRebalanceConfig);
+        String ioThreadName = NETWORK_THREAD_PREFIX + " | " + getClientId(config);
+        this.ioThread = new KafkaThread(ioThreadName, this.defaultBackgroundThread, true);
+        this.ioThread.start();
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: uhm always tricky to start threads in the constructor. Probably sensible to move this to the end of the constructor at least to avoid weird behaviors with initialization order

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 8]
```diff
@@ -233,7 +235,7 @@ private class OffsetFetchRequestState extends RequestState {
         public OffsetFetchRequestState(final Set<TopicPartition> partitions,
                                        final GroupState.Generation generation,
                                        final long retryBackoffMs) {
-            super(retryBackoffMs);
+            super(logContext, retryBackoffMs);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: I was thinking, to be consistent, maybe we should just pass a config in like other request manager, instead of parsing for a specific config.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 9]
```diff
@@ -555,10 +556,11 @@ public void close() throws IOException {
      * Check if the code is disconnected and unavailable for immediate reconnection (i.e. if it is in
      * reconnect backoff window following the disconnect).
      */
+    @Override
     public boolean isUnavailable(Node node) {
         lock.lock();
         try {
-            return client.connectionFailed(node) && client.connectionDelay(node, time.milliseconds()) > 0;
+            return NetworkClientUtils.isUnavailable(client, node, time);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Can we move this to a separated PR? This seems to be a refactor.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 10]
```diff
@@ -567,12 +569,11 @@ public boolean isUnavailable(Node node) {
...
      */
-    public void maybeThrowAuthFailure(Node node) {
+    @Override
+    public void maybeThrowAuthenticationException(Node node) {
         lock.lock();
         try {
-            AuthenticationException exception = client.authenticationException(node);
-            if (exception != null)
-                throw exception;
+            NetworkClientUtils.maybeThrowAuthenticationException(client, node);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Same - let's do it in a refactor PR.

File: clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
+/*
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java]: let's move this to the refactor PR and use it in the subsequent CTR PR.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java]: but awesome util!

**Review by philipnee — CHANGES_REQUESTED** [review]
Hey Kirk, Thanks for the PR.

Would it be possible to split this PR into 2-3 different PRs? Just to make it easier to cherry-pick in the future.  My recommendation is
1. Refactor PR: Include the refactors of the existing classes, idempotent closer, fetch collector, fetch utils, fetcher, and tests.
2. Refactor PR #2: You refactored some of the existing classes, maybe we can put this in a separated PR.
3. Maybe we can do 2 smaller PRs for this, like 1. Fetch Request Utils and 2. Fetch Request Manager and integration.

WDYT?
5. Fetcher

**Review by philipnee — DISMISSED** [review]
Hey Kirk, Thanks for the PR.

Would it be possible to split this PR into 2-3 different PRs? Just to make it easier to cherry-pick in the future.  My recommendation is
1. Refactor PR: Include the refactors of the existing classes, idempotent closer, fetch collector, fetch utils, fetcher, and tests.
2. Refactor PR 2: You refactored some of the existing classes, maybe we can put this in a separated PR.
3. Maybe we can do 2 smaller PRs for this, like 1. Fetch Request Utils and 2. Fetch Request Manager and integration.

WDYT?
5. Fetcher

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java [hunk 12]
```diff
@@ -127,19 +188,73 @@ private boolean doSend(final UnsentRequest r,
...
+
+        return list;
+    }
+
+    private void checkDisconnects() {
+        // any disconnects affecting requests that have already been transmitted will be handled
+        // by NetworkClient, so we just need to check whether connections for any of the unsent
+        // requests have been disconnected; if they have, then we complete the corresponding future
+        // and set the disconnect flag in the ClientResponse
+        for (Node node : unsentRequestNodes()) {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: The original implementation was not to check each individual node for connection as it is not necessary.  Instead, we wanted to just check if the request can still be connected to the existing node.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java [hunk 13]
```diff
@@ -127,19 +188,73 @@ private boolean doSend(final UnsentRequest r,
...
+                // coordinator failures traversing the unsent list again.
+                for (UnsentRequest unsentRequest : removeUnsentRequestByNode(node)) {
+                    // TODO: this should likely emulate what's done in ConsumerNetworkClient
+                    log.error("checkDisconnects - please update! unsentRequest: {}", unsentRequest);
+                }
             }
         }
     }
 
+    private void handlePendingDisconnects() {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: This is not needed because we are handling commit differently now. Here's the copy of the convo with Jason

Hey Jason, a couple questions here: there’s a bunch of queues in the consumerNetworkClient, in particular, unsent and pendingDisconnect: Is it something we want to re-implement in the new impl? Or we want to perform send and disconnect synchronously.


jason
:face_with_thermometer:  [12:59 PM](https://confluent.slack.com/archives/D03AZB2AXR7/p1668632389337969)
we discussed the unsent queue a little bit. the main reason to have this is for offset commits i think. as long as we have a way to expire an offset commit that could not be sent to the coordinator, i doubt we need to reimplement it
[1:00](https://confluent.slack.com/archives/D03AZB2AXR7/p1668632450426679)
i am not sure whether we need pending disconnect yet. i would say leave it out until we find a situation where we need it or something like it


pnee
:pizza-spin:  [1:03 PM](https://confluent.slack.com/archives/D03AZB2AXR7/p1668632633954739)
thanks, asking again because it seems like to puts all request into the unsent queue, so i don’t know if it impacts the performaces if we start to send things out synchronously…


jason
:face_with_thermometer:  [1:04 PM](https://confluent.slack.com/archives/D03AZB2AXR7/p1668632669510519)
we just need some way to track stuff that needs to be sent but cannot be sent now


pnee
:pizza-spin:  [1:04 PM](https://confluent.slack.com/archives/D03AZB2AXR7/p1668632689884719)
sounds good thanks

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: Ok I see the main usage of this is to try to disconnect async in the fetcher test.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 14]
```diff
@@ -0,0 +1,3510 @@
...
+        client.setWakeupHook(() -> {
+            if (!wokenUp.getAndSet(true)) {
+                networkClientDelegate.disconnectAsync(readReplica);
+                networkClientDelegate.poll(time.timer(0));
+            }
+        });
+
+        assertEquals(1, sendFetches());
+
+        networkClientDelegate.disconnectAsync(readReplica);
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: to the disconnectAsync - I wonder if we can deprecate this test (and therefore deprecate the disconnectAsync) in the refactor because the scenario might not be applicable anymore.  The reason this test was added I think is to handle multi-threading disconnection (from HB thread), which causes the fetcher to hang.  This is the original PR:

https://github.com/apache/kafka/pull/7511

and the description: 

This may be caused by the heartbeat thread concurrently handling a disconnection event just after the fetcher thread submits a request which would cause the Fetcher to enter an invalid state where it believes it has ongoing requests to the disconnected node but in fact it does not.

**Review by philipnee — COMMENTED** [review]
We might not need to change the network client delegate to the older state as a lot of functions there were added for legacy purpose.

**Review by philipnee — COMMENTED** [review]
We might not need to change the network client delegate to the older state as a lot of functions there were added for legacy purpose.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java (hunk 2 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: This method has been removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java (hunk 3 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: Indentation reverted.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java (hunk 5 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java]: Fixed

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java (hunk 6 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java]: Updated the description for `maybeThrowAuthFailure`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java (hunk 7 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java]: I have moved the thread starting logic elsewhere, but we should discuss.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 8 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: This code has been sufficiently changed that I'm not sure I understand the comment 😄

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java (hunk 9 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: I will do that. Thanks.

File: clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java (hunk 11 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java]: Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java (hunk 12 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: I have reverted this change in the `NetworkClientDelegate` and moved it to a subclass that's specific for testing needs.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java (hunk 13 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java]: Yes, I've moved this logic to the subclass of `NetworkClientDelegate` that is in the `FetchRequestManagerTest` class.

**Comment by kirktrue** [comment]
Re-draft-ing this PR in favor of PR #19 first.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 15]
```diff
@@ -0,0 +1,3630 @@
...
+    private FetchRequestManager<byte[], byte[]> fetchRequestManager;
+    private FetchCollector<byte[], byte[]> fetchCollector;
+    private TestableNetworkClientDelegate networkClientDelegate;
+    private OffsetFetcher offsetFetcher;
+
+    private MemoryRecords records;
+    private MemoryRecords nextRecords;
+    private MemoryRecords emptyRecords;
+    private MemoryRecords partialRecords;
+    private ExecutorService executorService;
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: not used?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 16]
```diff
@@ -0,0 +1,3630 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class FetchRequestManagerTest {
+
+    private static final double EPSILON = 0.0001;
+
+    private ConsumerRebalanceListener listener = new NoOpConsumerRebalanceListener();
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:150]: this should be declared final (along with the similar effectively-final declarations that follow)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 17]
```diff
@@ -0,0 +1,3630 @@
...
+
+    @Test
+    public void testMissingLeaderEpochInRecords() {
+        buildFetcher();
+
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
+
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, RecordBatch.MAGIC_VALUE_V0,
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:678]: this is autoclosable so I would create it in a try clause to ensure proper close

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 18]
```diff
@@ -0,0 +1,3630 @@
...
+
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
+
+        assertEquals(1, sendFetches());
+        client.prepareResponse(fullFetchResponse(tidp0, this.records, Errors.NONE, 100L, 0), true);
+        networkClientDelegate.poll(time.timer(0));
+        assertEmptyFetch("Should not return records or advance position on disconnect");
+
+        // disconnects should have no affect on subscription state
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:1897]: typo 'effect'

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 19]
```diff
@@ -0,0 +1,3630 @@
...
+    private ConsumerMetadata metadata;
+    private FetchMetricsRegistry metricsRegistry;
+    private FetchMetricsManager metricsManager;
+    private MockClient client;
+    private Metrics metrics;
+    private final ApiVersions apiVersions = new ApiVersions();
+    private ConsumerNetworkClient consumerClient;
+    private FetchRequestManager<byte[], byte[]> fetchRequestManager;
+    private FetchCollector<byte[], byte[]> fetchCollector;
+    private TestableNetworkClientDelegate networkClientDelegate;
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: I'm still trying to fully understand the specifics of the fetch manager related to this, but after a first pass I wonder why we need/want to involve a network client for unit testing the fetch manager... We should aim for a separation of the components, and test here the fetch manager logic without involving the network client. We've been doing it like that for the other managers, where we have been able to test the managers logic for building the requests, and then for processing the client response received (abstracting all the magic that the network client would do under the covers)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java (hunk 15 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: Removed the unused `executorService`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java (hunk 16 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:150]: `final`-ized many of the variables.

This file is largely copy and paste from the `FetcherTest` which suffers from the same indiscretions.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java (hunk 18 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:1897]: De-typo-ed.

This file is largely copy and paste from the `FetcherTest` which has the same typo.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java (hunk 19 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: @lianetm You're right, the `FetchRequestManagerTest` is unique among the request managers in that it uses the `NetworkClientDelegate`. The other request manager test classes rely on `mock`-ing and/or manually creating the `ClientResponse` to determine correctness. That's definitely a useful approach, but the current approach for the `FetchRequestManagerTest` has merit for the following reasons:

1. It doesn't "rock the boat"—`FetcherTest` uses the `ConsumerNetworkClient` throughout the tests and `FetchRequestManagerTest` is lifted from `FetcherTest` so it inherits those tests. The unit tests from `FetchTest` are exhaustive in their coverage, so I'm reluctant to rework them to the point where the use of the `NetworkClientDelegate` could be removed. (Note that the `NetworkClientDelegate` here wraps a `MockClient`, so it's not exercising the low-level network path.) In addition, the goal—once the old `Fetcher` is removed—is to replace `FetcherTest` with `FetchRequestManagerTest`. So keeping the code in the new test as close to the original as reasonably possible ensures the test coverage is comparable and hopefully makes the merge process less onerous.
2. `mock` can mask bugs—During development, I tried to integrate my code into `DefaultBackgroundThreadTest`, `DefaultEventHandlerTest`, and `PrototypeAsyncConsumerTest`. However, these tests were so heavily `mock`-ed that they neglected to exercise basic functionality of their respective companion classes. (This lack of coverage was borne out by code coverage tool results.) I uncovered a painful amount of silly errors in the underlying code, which took a while to resolve.
3. Testing the fetcher is miserable—The fetching logic is obstinately intertwined and spread across multiple classes. It could be argued that the `FetchRequestManagerTest` is closer to an integration test than a true unit test. The fetch process is pretty complicated and spans `AbstractFetch`, `CompletedFetch`, `Fetch`, `FetchBuffer`, `FetchCollector`, `FetchRequestManager`, and maybe even a couple of more. Without incorporating most of those components in the tests, the resulting tests would be anemically inferior to what was in `FetchTest`.

I had originally thought to have a second `FetchRequestManagerTest` that was simpler and relied on `mock`s, but ultimately I felt that the tests provided by the `FetcherTest` covered those anyway 🤷‍♂️

Thoughts?

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: I totally see how we are kind of inhering this, but my point is that it looks like the networkClient is mainly involved for preparing a client response, and processing it with poll (which will ultimately pass it to the manager that we want to test), so I'm trying to see the value it brings passing all requests/responses through that network client here. 

If we just poll the manager to build the requests (no network client involved), and then instead of:

```java
networkClientDelegate.addAll(pollResult.unsentRequests);
client.prepareResponse(fullFetchResponse(tidp0, this.records, Errors.NONE, 100L, 0));
networkClientDelegate.poll(time.timer(0));
``` 

We just provide the manager with the client response like:

```java
FetchResponse fetchResponse = fullFetchResponse(tidp0, this.records, Errors.NONE, 100L, 0);
ClientResponse clientResponse = new ClientResponse(
                new RequestHeader(ApiKeys.FETCH, (short) 1, "", 1),
                ...,
                fetchResponse);
pollResult.unsentRequests.get(0).future().complete(clientResponse);
``` 

With the above change we would be testing the manager functionality that we want, no need to involve a network client component. 
As an exercise I tried these changes in a couple of tests and works fine, and I do see lots with the same pattern of using the networkclient just to pass the request/response, but I totally get that this could still get to a point where is related to the network client in some other way that I might be missing

Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: I've started working to convert to this style now.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 20]
```diff
@@ -361,11 +354,14 @@ public void testFetchWithNoTopicId() {
...
 
-        client.prepareResponse(
-            fetchRequestMatcher((short) 12, noId, 0, Optional.of(validLeaderEpoch)),
+        completeFetches(
+            fetchRequest,
+            fetchRequestMatcher(version, noId, 0, Optional.of(validLeaderEpoch)),
+            version,
             fullFetchResponse(noId, this.records, Errors.NONE, 100L, 0)
         );
         networkClientDelegate.poll(time.timer(0));
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: This is not needed here anymore I expect

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 21]
```diff
@@ -468,6 +467,7 @@ public void testFetchForgetTopicIdWhenUnassigned() {
                 ),
                 singletonList(foo)
             ),
+            ApiKeys.FETCH.latestVersion(),
             fullFetchResponse(1, bar, this.records, Errors.NONE, 100L, 0)
         );
         networkClientDelegate.poll(time.timer(0));
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: Not needed anymore...

**Comment by lianetm** [comment]
Nice improvement removing the dependency on the networkClient when not needed in the unit tests. I think we can remove several calls to networkClientDelegate.poll that are not needed anymore, like the ones I pointed out above.

**Review by lianetm — APPROVED** [review]
Thanks for the changes! Just minor comment above to address

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java (hunk 19 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: OK, I can't make the transition to the simplified approach 😢

Around 85% of the methods can use the suggested approach, but there are about a dozen tests for which it won't work. Those remaining tests are exercising corner case behavior related to the interplay between the client and the fetcher for things like disconnects, timeouts, metadata changes, wakeup, etc. Any attempts I have made require a good amount of spaghetti code scaffolding that ends up duplicating functionality from `MockClient` anyway 🤷‍♂️ 

'tis no exaggeration—I have spent around 60 hours trying to reuse `FetcherTest` for `FetchRequestManager`. I was able to get the tests working nearly a month ago, and since then have been attempting to refactor to make things simpler and reduce diffs between my changes and `ctr-staging` and AK `trunk`.

At this point, the fetch request manager code is very well tested on account of leveraging `FetcherTest`. I would like to propose that we make a note of the need for future refactoring, use the `FetchRequestManagerTest` in its current, imperfect form, and move on.

WDYT?
